Support Flash HP driver for Renesas RA6, RA4 devices

khoa-nguyen-18 opened the pull request:
Add support for Flash HP driver for RA6, RA4 devices

- Add Flash HP driver support on EK_RA6E2, EK_RA6M1, EK_RA6M2, EK_RA6M3, EK_RA6M4, EK_RA6M5, FPB_RA6E1, FPB_RA6E2, EK_RA4E2, EK_RA4M2, EK_RA4M3
- Move Flash RA8 HW depend macro in flash_hp_ra.h to device tree configuration 
- Change to use irq_lock instead of seemaphore for code flash
- Modify and add conditions to check, erase and write at the last block
- Remove the `.code_in_ram` section in the linker file of all boards supported

khoa-nguyen-18 committed:
drivers: flash: update source code Flash driver for Renesas RA

- Bring macro defined of RA8 in flash_hp_ra.h to device tree
- Change to use irq_lock instead of semaphore for code flash
- Modify and add conditions to check and make decision to perform
action at last block.

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

khoa-nguyen-18 commented:
- Rebase main to solve the conflicts
- Add missing doc for flash for RA4E2, RA4M2, RA4M3

khoa-nguyen-18 committed:
dts: arm: renesas: Add Flash HP support for Renesas RA6, RA4

- Add Flash HP support for ra6-cm4, ra6-cm33, ra4-cm33 (except
r7fa4w1ad2cng)
- Add config to set the minimal size of data which can be written
for RA4E2, RA4M2, RA4M3, RA6E1, RA6E2, RA6M1, RA6M2, RA6M3, RA6M4,
RA6M5

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>

khoa-nguyen-18 committed:
boards: renesas: Add Flash HP support for Renesas RA6, RA4

- Add Flash HP support for ek_ra6m1, ek_ra6m2, ek_ra6m3, ek_ra6m4,
ek_ra6m5, ek_ra6e2, fpb_ra6e1, fpb_ra6e2, ek_ra4e2, ek_ra4m2,
ek_ra4m3
- Add doc about Flash for these boards

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>

khoa-nguyen-18 committed:
tests: drivers: flash: Support test common for RA6, RA4

Add overlay to support test flash/commone for: ek_ra6m1,
ek_ra6m2, ek_ra6m3, ek_ra6m4, ek_ra6m5, ek_ra6e2, fpb_ra6e1,
fpb_ra6e2, ek_ra4m2, ek_ra4m3, ek_ra4e2

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>

khoa-nguyen-18 commented:
Change the layout and modify the message of commits

khoa-nguyen-18 commented:
I just rebased main to solve conflicts and updated the revision for hal_renesas in `manifest` commit. @KhiemNguyenT @soburi @de-nordic can you please take a look?

thaoluonguw commented on dts/bindings/flash_controller/renesas,ra-flash-hp-controller.yaml:
```diff
@@ -6,3 +6,17 @@ description: Renesas RA family flash high-performance controller
 include: flash-controller.yaml
+
+properties:
+  block-32kb-linear-end:
+    type: int
+    required: true
+
+  block-32kb-dual-low-end:
+    type: int
+
+  block-32kb-dual-high-end:
+    type: int
+
+  reserved-area-num:
+    type: int
```
@khoa-nguyen-18 : Could you please add description for these properties?

khoa-nguyen-18 replied:
I have just added the description for these properties, thank you. Can you please take a look?

thaoluonguw commented on tests/drivers/flash/common/boards/ek_ra4e2.overlay:
```diff
@@ -0,0 +1,20 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* Set the 2 last block of storage. */
```
@khoa-nguyen-18 : Is it mistake? If it is correct, please check similar point.

khoa-nguyen-18 replied:
I'm sure this information is correct with the hardware manual, I also checked with other boards. Thank you, can you please take a look?

thaoluonguw replied:
Confirmed that comment updated. Thanks @khoa-nguyen-18 for your update.

khoa-nguyen-18 commented:
Hi, @de-nordic, @KhiemNguyenT, @soburi, @thaoluonguw, @duynguyenxa Can you please take a look at this Flash PR?

khoa-nguyen-18 commented:
rebase main to solve conflicts

khoa-nguyen-18 commented:
I just rebased main to solve the conflicts. @de-nordic @rghaddab @soburi @KhiemNguyenT, can you please take a look at this PR?

KhiemNguyenT commented:
@khoa-nguyen-18  please resolve the conflicts.

de-nordic commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -75,10 +47,29 @@ void bgo_callback(flash_callback_args_t *p_args)
+		if ((offset < 0) || (offset >= FLASH_HP_CF_DUAL_HIGH_END_ADDRESS) ||
+		    (offset >= FLASH_HP_CF_DUAL_LOW_END_ADDRESS &&
+		     offset < FLASH_HP_BANK2_OFFSET) ||
+		    ((len + offset) > FLASH_HP_CF_DUAL_HIGH_END_ADDRESS) ||
+		    ((len + offset) > FLASH_HP_CF_DUAL_LOW_END_ADDRESS &&
+		     (len + offset) < FLASH_HP_BANK2_OFFSET)) {
```
I am not sure whether you're completely integer overflow dafe here. Large len and offset can potentially overflow, as sum, to pass the checks.

khoa-nguyen-18 replied:
Thank you for kindly reminding me about this overflow case. I’ve updated it.

de-nordic commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -75,10 +47,29 @@ void bgo_callback(flash_callback_args_t *p_args)
+		}
+#else
+		if ((offset < 0) || (offset >= flash_data->area_size) ||
+		    (flash_data->area_size < len + offset)) {
```
Since you have already checked offset to be in range of area_size it would be better to do the comparison as 'flash_data->area_size - offset < len'.

khoa-nguyen-18 replied:
Thank you. I’ve updated it as well.

khoa-nguyen-18 committed:
soc: renesas: ra: Remove code_in_ram section

Remove code_in_ram section which is defined in sections.ld
of ra8m1, ra8d1, ra8t1, ra6m5, ra6m4, ra6m3, ra6m2, ra6m1,
ra6e2, ra6e1, ra4m3, ra4m2, ra4e2

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
manifest: Update hal_renesas for RA flash hp support

Update hal_renesas to support flash hp on RA

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented:
- Add conditions to check for overflow when summing 2 variables
- Remove the `.code_in_ram` section of all devices using `flash-hp`
- Add a commit to modify the revision of `hal_renesas`

khoa-nguyen-18 commented:
Rebase main to solve conflicts

khoa-nguyen-18 commented:
Add support [flash_get_size](https://docs.zephyrproject.org/latest/doxygen/html/group__flash__interface.html#ga471edb53a2b1abe3dff0f4e2d216521e) api

khoa-nguyen-18 commented:
@de-nordic can you please have a look at this PR?
